pkg/etcdcli/health.go

```diff
@@ -0,0 +1,104 @@
+package etcdcli
+
+import (
+	"fmt"
+	"strings"
+
+	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+)
+
+type healthCheck struct {
+	Member  *etcdserverpb.Member
+	Health  bool
+	Started bool
+	Took    string
+	Error   error
+}
+
+type memberHealth struct {
+	Check []healthCheck
+}
+
+// Status returns a reporting of memberHealth status
+func (h *memberHealth) Status() string {
+	etcdMembers := h.Check
+	healthyMembers := h.GetHealthyMemberNames()
+
+	status := []string{}
+	if len(etcdMembers) == len(healthyMembers) {
+		status = append(status, fmt.Sprintf("%d members are available", len(etcdMembers)))
+	} else {
+		status = append(status, fmt.Sprintf("%d of %d members are available", len(healthyMembers), len(etcdMembers)))
+		for _, etcd := range h.Check {
+			switch {
+			case !etcd.Started:
+				status = append(status, fmt.Sprintf("%s has not started", etcd.Member.Name))
+				continue
+			case !etcd.Health:
+				status = append(status, fmt.Sprintf("%s is unhealthy", etcd.Member.Name))
+				continue
+			}
+		}
+	}
+	return strings.Join(status, ", ")
+}
+
+// Healthy returns healthy members
+func (h *memberHealth) GetHealthyMembers() []*etcdserverpb.Member {
+	members := []*etcdserverpb.Member{}
+	for _, etcd := range h.Check {
+		if etcd.Health {
+			members = append(members, etcd.Member)
+		}
+	}
+	return members
+}
+
+// HealthyMemberNames returns a list of healthy member names
+func (h *memberHealth) GetHealthyMemberNames() []string {
+	memberNames := []string{}
+	for _, etcdMember := range h.GetHealthyMembers() {
+		memberNames = append(memberNames, etcdMember.Name)
+	}
+	return memberNames
+}
+
+// Unhealthy returns unhealthy members
+func (h *memberHealth) GetUnhealthyMembers() []*etcdserverpb.Member {
```

Comment:
It seems this method and GetHealthyMembers() can be combined into one which accepts a bool flag (healthy member or not).

Reply:
In this case I personally prefer explicit functions that do one thing than introducing modality through a boolean.

Reply:
Mainly because you can extend the combination argument to pretty much every function here (e.g. why not one function with a filter argument which could allow every filter permutation to be expressed).